src/rust/inetstack/protocols/tcp/tests/simulator.rs

```diff
@@ -850,36 +850,48 @@ impl Simulation {
     }
 
     /// Builds a TCP segment.
-    fn build_tcp_segment(&self, tcp_packet: &TcpPacket) -> TcpSegment {
+    fn build_tcp_segment(&self, tcp_packet: &TcpPacket) -> DemiBuffer {
         // Create headers.
-        let ethernet2_hdr: Ethernet2Header = self.build_ethernet_header();
         let ipv4_hdr: Ipv4Header = self.build_ipv4_header(IpProtocol::TCP);
         let tcp_hdr: TcpHeader = self.build_tcp_header(&tcp_packet);
         let data: Option<DemiBuffer> = if tcp_packet.seqnum.win > 0 {
             Some(Self::cook_buffer(tcp_packet.seqnum.win as usize, None))
         } else {
             None
         };
+        let mut segment: TcpSegment =
+            TcpSegment::new(ipv4_hdr, tcp_hdr, data, false).expect("Should be able to create TCP segment");
+        let mut pkt: DemiBuffer = segment.take_body().unwrap();
 
-        TcpSegment::new(ethernet2_hdr, ipv4_hdr, tcp_hdr, data, false).unwrap()
+        // Actually prepend L2 header.
```

Comment:
This is identical for tcp and udp, we should extract it into a separate fn if possible.